File: time/model/parse.py
import os
import pandas as pd
import random
from tqdm import tqdm

DIR = "_processed"

TRAINING_RATIO = 0.95

TRAINING_CHUNKS = 14
VALIDATION_CHUNKS = 1


HEADERS = "fen,isDraw,isWin,ply,totalPly,qSearch,inCheck,increment,timeLeft,timeSpent,totalTimeSpent,startTime,opponentTime"


def write_headers():
    for chunk in range(TRAINING_CHUNKS):
        with open(f"training/chunk_{chunk}.csv", "w") as f:
            f.write(HEADERS + "\n")

    for chunk in range(VALIDATION_CHUNKS):
        with open(f"validation/chunk_{chunk}.csv", "w") as f:
            f.write(HEADERS + "\n")


def count_rows(csv):
    chunk_size = 256 * 1024 * 1024
    total = 0
    with open(csv, "rb") as f:
        while chunk := f.read(chunk_size):
            total += chunk.count(b"\n")
    return total - 1


def main():
    for d in ["training", "validation"]:
        os.makedirs(d, exist_ok=True)

    write_headers()

    training_writers = [
        open(f"training/chunk_{i}.csv", "a") for i in range(TRAINING_CHUNKS)
    ]
    validation_writers = [
        open(f"validation/chunk_{i}.csv", "a") for i in range(VALIDATION_CHUNKS)
    ]

    for file in [x.path for x in os.scandir(DIR) if x.is_file()]:
        total_rows = count_rows(file)
        chunksize = 100000
        total_chunks = total_rows // chunksize + bool(total_rows % chunksize)
        reader = pd.read_csv(file, chunksize=chunksize)
        for chunk in tqdm(
            reader, total=total_chunks, desc=os.path.basename(file), leave=True
        ):
            rows = chunk.to_csv(index=False, header=False).splitlines()
            for row in rows:
                if random.random() < TRAINING_RATIO:
                    idx = random.randint(0, TRAINING_CHUNKS - 1)
                    training_writers[idx].write(row + "\n")
                else:
                    idx = random.randint(0, VALIDATION_CHUNKS - 1)
                    validation_writers[idx].write(row + "\n")
        os.remove(file)

    for writer in training_writers + validation_writers:
        writer.close()


if __name__ == "__main__":
    main()

File: time/dataset/parse.cpp
#include <filesystem>
#include <sstream>
#include "../../pipeline/dataparser.h"
#include "../../pipeline/utils.h"
#include "../include/datum.h"

inline Datum parseLine(const std::string& line) {
    Datum res;
    std::istringstream lineStream(line);
    std::string s;
    size_t i = 0;
    while (std::getline(lineStream, s, ',')) {
        switch (i) {
        case 0: {
            std::istringstream fenStream(s);
            std::string fenPart;
            size_t j = 0;
            while (j < 2 && std::getline(fenStream, fenPart, ' ')) {
                switch (j) {
                case 0:
                    parseFen(fenPart, res.pos);
                    break;
                case 1:
                    res.side = fenPart != "w";
                    break;
                }
                ++j;
            }
            break;
        }
        case 1:
            res.isDraw = s == "1";
            break;
        case 2:
            res.isWin = s == "1";
            break;
        case 3:
            res.ply = std::stoi(s);
            break;
        case 4:
            res.totalPly = std::stoi(s);
            break;
        case 5:
            res.qSearch = std::stoi(s);
            break;
        case 6:
            res.inCheck = s == "1";
            break;
        case 7:
            res.increment = std::stoi(s);
            break;
        case 8:
            res.timeLeft = std::stoi(s);
            break;
        case 9:
            res.timeSpent = std::stoi(s);
            break;
        case 10:
            res.totalTimeSpent = std::stoi(s);
            break;
        case 11:
            res.startTime = std::stoi(s);
            break;
        case 12:
            res.opponentTime = std::stoi(s);
            break;
        }
        ++i;
    }
    return res;
}

int main() {
    std::filesystem::path cwd = std::filesystem::current_path();
    DataParser<6, 25000000, 0.95, Datum, parseLine> dataParser(
        cwd / "_processed", ".csv", cwd / "training", cwd / "validation");
    dataParser.parseFiles();

    return 0;
}
